prawns/balances/src/processor.ts

```diff
@@ -15,15 +24,202 @@ if (!supportedNetworks.includes(network)) {
   throw Error('Network not supported');
 }
 
-const processor = new SubstrateProcessor(new TypeormDatabase())
+export const processor = new SubstrateBatchProcessor()
   .setBatchSize(500)
   .setTypesBundle(network)
   .setDataSource({
     archive: lookupArchive(network as KnownArchives, { release: 'FireSquid' }),
   })
-  .addEventHandler('Balances.Transfer', balanceTransferHandler(network, 0))
-  .addEventHandler('Balances.BalanceSet', balanceSetHandler(network, 0))
-  .addEventHandler('Balances.Endowed', balanceEndowedHandler(network, 0))
-  .addEventHandler('Balances.Deposit', balanceDepositHandler(network, 0))
-  .addEventHandler('Treasury.Awarded', treasuryAwardedHandler(network, 0))
-  .run();
+  .addEvent('Balances.Transfer', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.BalanceSet', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.Endowed', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Balances.Deposit', {
+    data: { event: { args: true } },
+  } as const)
+  .addEvent('Treasury.Awarded', {
+    data: { event: { args: true } },
+  } as const);
+
+// doing .run above breaks the types... bit odd
+processor.run(new TypeormDatabase(), processBatch);
+
+async function processBatch(ctx: Context): Promise<void> {
+  const models: Models = {
+    accounts: [],
+    changeEvents: [],
+    transfers: [],
+  };
+
+  // synchronously process all data in the batch, mutating models (above) as the batch's data cache
+  for (const block of ctx.blocks) {
+    for (const item of block.items) {
+      if (item.kind === 'call') {
+        // we have none, leaving for reference
+      } else if (item.kind === 'event') {
+        const blockNumber = BigInt(block.header.height);
+        const date = new Date(block.header.timestamp);
+        processEvent({
+          ctx,
+          item,
+          blockNumber,
+          date,
+          models,
+          network,
+        });
+      }
+    }
+  }
+
+  // save accounts first as they are FKeyed in the other models
+  const existingAccounts = await ctx.store.findBy<SubstrateBalanceAccount>(
+    SubstrateBalanceAccount,
+    { id: In(models.accounts.map((acc) => acc.id)) }
+  );
+  await ctx.store.save(mergeAccounts(models.accounts, existingAccounts));
+
+  await Promise.all([
+    ctx.store.save(models.transfers),
```

Comment:
I think save takes an array... so I think you can do a single save with all models

Reply:
that is an array, but I was a bit nervous about sending an array of different model types! I'll give it a go

Reply:
It definitely worked pre-firesquid as I did it in governance somewhere.

Reply:
Updated, I've left accounts above as it's essential they go in first